src/storage-types/src/read_holds.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fmt::Debug;
+
+use mz_repr::GlobalId;
+use timely::progress::{Antichain, ChangeBatch, Timestamp as TimelyTimestamp};
+use timely::PartialOrder;
+use tokio::sync::mpsc::UnboundedSender;
+
+/// Token that represents a hold on a collection. This prevents the since of the
+/// collection from progressing beyond the hold. In other words, it cannot
+/// become true that our hold is `less_than` the since.
+///
+/// This [ReadHold] is safe to drop. The installed read hold will be returned to
+/// the issuer behind the scenes.
+pub struct ReadHold<T: TimelyTimestamp> {
+    /// Identifies that collection that we have a hold on.
+    pub id: GlobalId,
+
+    /// The times at which we hold.
+    pub since: Antichain<T>,
+
+    /// For communicating changes to this read hold back to whoever issued it.
+    pub holds_tx: UnboundedSender<(GlobalId, ChangeBatch<T>)>,
+}
+
+impl<T: TimelyTimestamp> Debug for ReadHold<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ReadHold")
+            .field("id", &self.id)
+            .field("since", &self.since)
+            .finish_non_exhaustive()
+    }
+}
+
+impl<T: TimelyTimestamp> ReadHold<T> {
+    pub fn new(
+        id: GlobalId,
+        since: Antichain<T>,
+        holds_tx: UnboundedSender<(GlobalId, ChangeBatch<T>)>,
+    ) -> Self {
+        Self {
+            id,
+            since,
+            holds_tx,
+        }
+    }
+
+    /// Returns the [GlobalId] of the collection that this [ReadHold] is for.
+    pub fn id(&self) -> &GlobalId {
+        &self.id
+    }
+
+    /// Returns the frontier at which this [ReadHold] is holding back the since
+    /// of the collection identified by `id`. This does not mean that the
+    /// overall since of the collection is what we report here. Only that it is
+    /// _at least_ held back to the reported frontier by this read hold.
+    pub fn since(&self) -> &Antichain<T> {
+        &self.since
+    }
+
+    /// Merges `other` into `self`, keeping the overall read hold.
+    pub fn merge_assign(&mut self, mut other: ReadHold<T>) {
```

Comment:
The semantics of this method are a bit unclear. What happens if I give it two read holds for different `id`s? Looking at the implementation it looks like we will treat `other` like it is a read hold for the same collection as `self` and then we might end up with (a) a frontier regression for `self.id` and (b) a leaked read hold for `other.id`? Also weird things will happen when you get two read holds with different `holds_tx`s.

I think you'd at least want to assert that the two IDs and channels are equal. And then, isn't the implementation equivalent to just dropping the `ReadHold` with the greater `since`?

Reply:
Oh boy, good catch! This code is quite old by now, and has remnants from a time when I had a `ReadHold` that contained multiple IDs. Will fix!

> I think you'd at least want to assert that the two IDs and channels are equal. And then, isn't the implementation equivalent to just dropping the ReadHold with the greater since?

With our timestamp yes, but because the timestamp type is generic here we could have more complicated `sinces` with partial orderings.

Reply:
Ah, good point!